openshift-kube-apiserver/admission/customresourcevalidation/dns/validate_dns.go

```diff
@@ -0,0 +1,203 @@
+	if uncastObj == nil {
+		return nil, nil
+	}
+
+	obj, ok := uncastObj.(*operatorv1.DNS)
+	if !ok {
+		return nil, field.ErrorList{
+			field.NotSupported(field.NewPath("kind"), fmt.Sprintf("%T", uncastObj), []string{"DNS"}),
+			field.NotSupported(field.NewPath("apiVersion"), fmt.Sprintf("%T", uncastObj), []string{"operator.openshift.io/v1"}),
+		}
+	}
+
+	return obj, nil
+}
+
+// dnsV1 is runtime object that is validated as a versioned DNS.
+type dnsV1 struct{}
+
+// ValidateCreate validates a DNS that is being created.
+func (dnsV1) ValidateCreate(uncastObj runtime.Object) field.ErrorList {
+	obj, errs := toDNSV1(uncastObj)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	errs = append(errs, validation.ValidateObjectMeta(&obj.ObjectMeta, false, validation.NameIsDNSSubdomain, field.NewPath("metadata"))...)
+	errs = append(errs, validateDNSSpecCreate(obj.Spec)...)
+
+	return errs
+}
+
+// ValidateUpdate validates a DNS that is being updated.
+func (dnsV1) ValidateUpdate(uncastObj runtime.Object, uncastOldObj runtime.Object) field.ErrorList {
+	obj, errs := toDNSV1(uncastObj)
+	if len(errs) > 0 {
+		return errs
+	}
+	oldObj, errs := toDNSV1(uncastOldObj)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	errs = append(errs, validation.ValidateObjectMetaUpdate(&obj.ObjectMeta, &oldObj.ObjectMeta, field.NewPath("metadata"))...)
+	errs = append(errs, validateDNSSpecUpdate(obj.Spec, oldObj.Spec)...)
+
+	return errs
+}
+
+// ValidateStatusUpdate validates a DNS status that is being updated.
+func (dnsV1) ValidateStatusUpdate(uncastObj runtime.Object, uncastOldObj runtime.Object) field.ErrorList {
+	obj, errs := toDNSV1(uncastObj)
+	if len(errs) > 0 {
+		return errs
+	}
+	oldObj, errs := toDNSV1(uncastOldObj)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	errs = append(errs, validation.ValidateObjectMetaUpdate(&obj.ObjectMeta, &oldObj.ObjectMeta, field.NewPath("metadata"))...)
+
+	return errs
+}
+
+// validateDNSSpecCreate validates the spec of a DNS that is being created.
+func validateDNSSpecCreate(spec operatorv1.DNSSpec) field.ErrorList {
+	return validateDNSSpec(spec)
+}
+
+// validateDNSSpecUpdate validates the spec of a DNS that is being updated.
+func validateDNSSpecUpdate(newspec, oldspec operatorv1.DNSSpec) field.ErrorList {
+	return validateDNSSpec(newspec)
+}
+
+// validateDNSSpec validates the spec of a DNS.
+func validateDNSSpec(spec operatorv1.DNSSpec) field.ErrorList {
+	var errs field.ErrorList
+	specField := field.NewPath("spec")
+	errs = append(errs, validateDNSNodePlacement(spec.NodePlacement, specField.Child("nodePlacement"))...)
+	return errs
+}
+
+// validateDNSSpec validates the spec.nodePlacement field of a DNS.
+func validateDNSNodePlacement(nodePlacement operatorv1.DNSNodePlacement, fldPath *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	if len(nodePlacement.NodeSelector) != 0 {
+		errs = append(errs, unversionedvalidation.ValidateLabels(nodePlacement.NodeSelector, fldPath.Child("nodeSelector"))...)
```

Comment:
just confirm with me that this field is new in 4.8

Reply:
Yes, the `spec.nodePlacement` field is new in the `dnses.operator.openshift.io` resource in 4.8.